### ISSUE TITLE
Usage with `pmap`

### ISSUE THREAD

**Issue by shyams2** [mpi4jax/mpi4jax#121]
Hi

Firstly, I have to say this is a wonderful library and thank you for open-sourcing it :)

Currently, I just have a few questions I'm hoping to find more clarity on. I'm trying to perform data-parallel training of neural networks using your library on GPUs. Till now, I've been using `pmap` to parallelize across GPUs with collective communications performed by `jax.lax.pmean`/`jax.lax.psum`. However, this only works on a single host. Now, if I were to move to a cluster which has say 4 GPUs per host, these are the code changes that I can anticipate:
- The batch would now need to be split accounting for the total number of processes in addition to the number of local devices. So something like `reshape(comm.Get_size() * jax.local_device_count, ...)`. Then using `batch = data[comm.Get_rank() * jax.local_device_count():(comm.Get_rank() + 1) * jax.local_device_count(), ...]
- The gradients would now need to be averaged first across all devices on each node and then averaged across the nodes. So we would have something like `grads_avg, _ = mpi4jax.allreduce(result_after_pmean, op=MPI.SUM, comm=comm); grads_avg = grads_avg / comm.Get_size()`

Would this be the correct approach? Are there any problems that can be expected when blending `pmap` and `mpi4jax`? Is there any example code of such a use-case I can potentially build on?

Thanks!

**Comment by PhilipVinc** [comment]
Hi @shyams2 , thanks for the nice words!

I'm not sure what would happen if you mix pmap-distributed arrays (among local devices) and MPI. 
We did not put in the work to support it, and i suspect it will fail (though, if you want to contribute to make it work, a PR would be well accepted!)

I have a quite large codebase working with multiple GPUs on multiple nodes (NetKet). 
The way we approach this is to assume that every MPI rank/process only works on/with a single GPU. Then you treat all GPUs as different processes through MPI and you don't worry about anything.

I generally do something like

run.py
```python
import os

# detect MPI rank
from mpi4py import MPI
rank = MPI.COMM_WORLD.Get_rank()

# set only one visible device
os.environ["CUDA_VISIBLE_DEVICES"] = f"{rank}"
# force to use gpu
os.environ["JAX_PLATFORM_NAME"] = "gpu"

# you must have set the ENV variables before loading jax!
import jax 
import netket as nk
import mpi4jax
from mpi4py import MPI

print(f"{rank} -> {jax.devices()}")

N_gpus_total = MPI.MPI_COMM_WORLD.Get_size()
local_rank = MPI.MPI_COMM_WORLD.Get_rank()

global_data = jnp.random.rand(jax.random.Token(0), (N_gpus_total, 1000))
local_data = global_data[local_rank, :]
local_transformed_data = local_data * 2
reduction = mpi4jax.allreduce(MPI.SUM, local_data)
```

```
mpirun run.py
```

**Comment by shyams2** [comment]
@PhilipVinc , thanks for the informative response! I'll can definitely just try out considering every GPU as a separate MPI process, and see how that fares. I have one question here:  If MPI is compiled with CUDA support as well, would communication between GPUs across hosts be zero-copy as well?

I tried a small toy example to see if `pmap` will actually play well with `mpi4jax`. Does seem like it works at least when I didn't enable CUDA-aware MPI (It segfaulted when I tried - however this may just be because I didn't compile it with CUDA support):
```python
from mpi4py import MPI
import jax
import jax.numpy as jnp
import mpi4jax
import functools

comm = MPI.COMM_WORLD
rank = comm.Get_rank()

@jax.pmap
def data_generator(rng):
    return jax.random.normal(rng, (10,))

@functools.partial(jax.pmap, axis_name = 'i')
def psum_data(data):
    return jax.lax.psum(data, axis_name = 'i')

@jax.jit
def mpi_test(arr):
   arr_sum, _ = mpi4jax.allreduce(arr, op=MPI.SUM, comm=comm)
   return arr_sum

rng_key = jax.random.PRNGKey(rank)
keys = jax.random.split(rng_key, jax.local_device_count())

out = data_generator(keys)
out_pmap= psum_data(out)
out_MPI= mpi_test(out_pmap)

print('Rank:', rank)
print('out', out)
print('out_pmap', out_pmap)
print('out_MPI', out_MPI)
print()
```

**Comment by shyams2** [comment]
I have another follow-up question regarding the use of pure-MPI for communication. So I tried doing the same toy example that I did above using pure MPI for communication. However, I'm seeing some difficulty in getting it working across several nodes - it seems to throw an OOM error for some reason. In my case, I'm working on nodes that have 4 GPUs, and I tried running it across 8 GPUs on 2 nodes. So I'm trying to pick the device by doing `rank % 4`. Would this be the right way to do so?:
```python
import os
from mpi4py import MPI

# detect MPI rank
rank = MPI.COMM_WORLD.Get_rank()
# finding the device ID:
device_id = rank % 4

# set only one visible device
os.environ["CUDA_VISIBLE_DEVICES"] = f"{device_id}"
# force to use gpu
os.environ["JAX_PLATFORM_NAME"] = "gpu"

import jax
import time
import jax.numpy as jnp
import mpi4jax
import functools

comm = MPI.COMM_WORLD

@jax.jit
def data_generator(rng):
    return jax.random.normal(rng, (100000,))

@jax.jit
def mpi_test(arr):
   arr_sum, _ = mpi4jax.allreduce(arr, op=MPI.SUM, comm=comm)
   return arr_sum

rng_key = jax.random.PRNGKey(rank)

out = data_generator(rng_key)
out_MPI= mpi_test(out)

tic = time.perf_counter()
for i in range(100):
    out = data_generator(keys)
out_time = time.perf_counter() - tic

tic = time.perf_counter()
for i in range(100):
    out_MPI = mpi_test(out)
out_MPI_time = time.perf_counter() - tic

print('Rank:', rank)
print('out', out_time)
print('out_MPI', out_MPI_time)
print()
```

**Comment by dionhaefner** [comment]
Seems right to me. How did you run this example? I suspect you didn't actually run on 2 nodes for some reason, so some ranks shared the same device which leads to OOM because of JAX' greedy preallocation.

**Comment by shyams2** [comment]
I'd submitted a slurm script to run the file. Do you think there are any issues with this?:
```
#!/bin/sh -l
# FILENAME: submit_script

#SBATCH --gpus-per-node=4
#SBATCH -p gpu
#SBATCH --nodes=2
#SBATCH --ntasks=8
#SBATCH --time=0:20:00
#SBATCH --output=/home/user/testing_dir/myjob.out
#SBATCH --error=/home/user/testing_dir/myjob_err.out

module load modtree/gpu
module load anaconda
conda activate

# Change to the directory from which you originally submitted this job.
cd $SLURM_SUBMIT_DIR

# Run the python code:
mpiexec -n 8 python3 pure_mpi_test.py
```

**Comment by shyams2** [comment]
So, it seemed like I needed some more options in the bash script I submitted to sbatch. I got it working after all - thank you for all the help!

I'm closing this issue now.

**Comment by dionhaefner** [comment]
Glad you figured it out. Could you post the solution for others running into the same problem?

**Comment by shyams2** [comment]
Sure thing. Wasn't sure if this was something cluster specific. I added on the following to get it working
```bash
#SBATCH --ntasks-per-node=4
#SBATCH --gpus-per-task=1
```

**Comment by dionhaefner** [comment]
And with `--gpus-per-task` you didn't need to modify `CUDA_VISIBLE_DEVICES`, right?

**Comment by shyams2** [comment]
I didn't remove that from my code, and it ran fine. Let me see if it works with that line removed then
